File: Days.34/Python/4.Median-in-a-Running-Stream.py
# Median in a Running Stream

import math, heapq

class Solution:
    def __init__(self):
        self.maxheap=[]
        self.minheap=[]
    
    def BalanceHeaps(self):
        if len(self.minheap)>len(self.maxheap):
            heapq.heappush(self.maxheap,-heapq.heappop(self.minheap))
        elif len(self.maxheap)>len(self.minheap)+1:
            heapq.heappush(self.minheap,-heapq.heappop(self.maxheap))
    def insertHeaps(self,x):
        if len(self.maxheap)==0 or x<-self.maxheap[0]:
            heapq.heappush(self.maxheap,-x)
        else:
            heapq.heappush(self.minheap,x)
        self.BalanceHeaps()

    def getMedian(self):
        if len(self.maxheap)>len(self.minheap):
            return -self.maxheap[0]
        else:
            return (-self.maxheap[0]+self.minheap[0])/2
        
def main():
    n=int(input())
    solution=Solution()
    for i in range(n):
        x=int(input())
        solution.insertHeaps(x)
        print(math.floor(solution.getMedian()))



main()

File: Days.34/C++/4.Median-in-a-Running-Stream.cpp
// Median in a Running Stream

#include<iostream>
#include<vector>
#include<queue>
using namespace std;

class Solution
{
    priority_queue<int,vector<int> > maxheap;
    priority_queue<int,vector<int>,greater<int> > minheap;
    public:
    void balanceHeaps()
    {
        int val;
        if(minheap.size()>maxheap.size())
        {
            val=minheap.top();
            minheap.pop();
            maxheap.push(val);
        }
        else if(maxheap.size()>minheap.size()+1)
        {
            val=maxheap.top();
            maxheap.pop();
            minheap.push(val);
        }
    }
    void insertHeap(int x)
    {
        if(maxheap.size()==0 or x<maxheap.top())
        {
            maxheap.push(x);
        }
        else
        {
            minheap.push(x);
        }
        balanceHeaps();
    }
    double getMedian()
    {
        if(maxheap.size()>minheap.size())
        {
            return maxheap.top();
        }
        else
        {
            return (maxheap.top()+minheap.top())/2.0;
        }
    }
};

int main()
{
    int n;
    cin>>n;
    int x;
    Solution solution;
    for(int i=0; i<n; i++)
    {
        cin>>x;
        solution.insertHeap(x);
        cout<<"Median: "<<solution.getMedian()<<"\n";
    }
    return 0;
}